app/forms/declaration/PackageInformation.scala

```diff
@@ -17,50 +17,84 @@
 package forms.declaration
 
 import forms.DeclarationPage
-import play.api.data.Forms.{number, text}
+import models.DeclarationType.{CLEARANCE, DeclarationType}
+import play.api.data.Forms.{number, optional, text}
 import play.api.data.{Form, Forms}
 import play.api.libs.json.{JsValue, Json}
 import services.PackageTypes
 import utils.validators.forms.FieldValidator._
 
-case class PackageInformation(typesOfPackages: String, numberOfPackages: Int, shippingMarks: String) {
+case class PackageInformation(typesOfPackages: Option[String], numberOfPackages: Option[Int], shippingMarks: Option[String]) {
   def toJson: JsValue = Json.toJson(this)(PackageInformation.format)
 
-  def typesOfPackagesText: String = PackageTypes.findByCode(typesOfPackages).asText()
+  def typesOfPackagesText: Option[String] =
+    for {
+      types <- typesOfPackages
+      packageType <- PackageTypes.findByCode(types)
+    } yield packageType.asText()
 }
 
 object PackageInformation extends DeclarationPage {
 
-  def fromJsonString(value: String): Option[PackageInformation] = Json.fromJson(Json.parse(value)).asOpt
+  private def fromValues(typesOfPackages: String, numberOfPackages: Int, shippingMarks: String): PackageInformation =
+    PackageInformation(typesOfPackages = Some(typesOfPackages), numberOfPackages = Some(numberOfPackages), shippingMarks = Some(shippingMarks))
+
+  private def toValues(packageInformation: PackageInformation): Option[(String, Int, String)] =
+    for {
+      typesOfPackages <- packageInformation.typesOfPackages
+      numberOfPackages <- packageInformation.numberOfPackages
```

Comment:
If typesOfPackages will be empty you skip rest of the values

Reply:
I will return `None`. To be fair, I am not sure how it should behave in that case:
1. Return `None`
2. Return some default values for the fields and `None` only if all are empty

Reply:
But for comprehensive it's combination of flatMaps and maps

```
packageInformation.typesOfPackages.flatMap { typesOfPackages: String =>
  packageInformation.numberOfPackages.flatMap { numberOfPackages: Int =>
    packageInformation.shippingMarks.map { shippingMarks: String =>
      (typesOfPackages, numberOfPackages, shippingMarks)
    }
  }
}
```

This is the equivalent of above code, in case that one of the values will be empty e.g. `typesOfPackages` method `toValues` will return `None`. Is it expected behaviour? Because I'm not sure what you want to achieve with this method when the `typeOfPackages` is empty.

In case if the above scenario is correct this method is fine, but I want to confirm this. I'm not sure if you can have a `numberOfPackages` as defined and rest not defined.

In general it can be my wrong understanding what this method should do.

Reply:
I see where are you using it. Can you test scenario when you're doing mandatory journey, where those fields are mandatory and you'll fill just the last input?

I'm not sure about behaviour too

Reply:
Hmm, it works good. Sorry for the confusion. I will discard my request changes to this. Still I want to understand what should be the behaviour of this method and when it will be invoked

Reply:
Can't this simply be
` = Some(packageInformation.type, packageInformation.number, packageInformation.shipping)`
Or am I missing something

Reply:
Unfortunatelly, it can't.
All fields in `PackageInformation` are `Option`s, so I could either provide some default values (which makes problem for `Int`, because '0' is a valid answer), or I could return `None` in case the object is in invalid state.
The latter represents the "fail fast" approach.

Reply:
Sorry, I read that method signature wrong.  Thought it was `Option[(Option[String], Option[Int], Option[String])]`